apis/Google.Cloud.Metadata.V1/Google.Cloud.Metadata.V1/MetadataClientImpl.cs

```diff
@@ -0,0 +1,702 @@
+        private const string MetadataFlavor = "Metadata-Flavor";
+        private const string GoogleMetadataHeader = "Google";
+
+        // TODO: Should this be created with a MetadataClient instead?
+        private static readonly ILogger Logger = ApplicationContext.Logger.ForType<MetadataClientImpl>();
+
+        private readonly string metadataHost;
+
+        private readonly object eventLock = new object();
+        private ValueWatcher instanceWatcher;
+        private ValueWatcher maintenanceStatusWatcher;
+        private ValueWatcher projectWatcher;
+
+        public override ConfigurableHttpClient HttpClient { get; }
+
+        public MetadataClientImpl(ConfigurableHttpClient httpClient = null)
+        {
+            HttpClient = httpClient ?? new HttpClientFactory().CreateHttpClient(new CreateHttpClientArgs());
+            metadataHost = GetMetadataHost();
+        }
+
+        public MetadataClientImpl(IHttpClientFactory httpClientFactory)
+        {
+            HttpClient = (httpClientFactory ?? new HttpClientFactory()).CreateHttpClient(new CreateHttpClientArgs());
+            metadataHost = GetMetadataHost();
+        }
+
+        private static MaintenanceStatus GetMaintenanceStatus(string result)
+        {
+            switch (result)
+            {
+                case "NONE": return MaintenanceStatus.None;
+                case "MIGRATE_ON_HOST_MAINTENANCE": return MaintenanceStatus.Migrate;
+                default: return MaintenanceStatus.Unknown;
+            }
+        }
+
+        private string GetMetadataHost()
+        {
+            var emulatorHost = Environment.GetEnvironmentVariable(EmulatorEnvironmentVariable);
+            return string.IsNullOrEmpty(emulatorHost) ? DefaultMetadataHost : emulatorHost;
+        }
+
+        private static T GetResult<T>(Func<Task<T>> operation)
+        {
+            try
+            {
+                return operation().Result;
+            }
+            catch (AggregateException e)
+            {
+                throw e.InnerExceptions.FirstOrDefault() ?? e;
+            }
+        }
+
+        private async Task<HttpResponseMessage> RequestMetadata(string relativeUrl, CancellationToken cancellationToken)
+        {
+            var httpRequest = new HttpRequestMessage(HttpMethod.Get, $"http://{metadataHost}/computeMetadata/v1/{relativeUrl}");
+            httpRequest.Headers.Add(MetadataFlavor, GoogleMetadataHeader);
+            return await HttpClient.SendAsync(httpRequest, cancellationToken).ConfigureAwait(false);
+        }
+
+        private static readonly Regex KeyRegex = new Regex(@"\A[a-zA-Z0-9-_]{1,127}\Z", RegexOptions.Compiled);
+        private static void ValidateCustomMetadataKey(string key)
+        {
+            if (!KeyRegex.IsMatch(key))
+            {
+                throw new ArgumentException("Invalid custom metadata key.", "key");
+            }
+        }
+
+        /// <summary>
```

Comment:
If this is a copy of the doc comment in the base class, I'd use `/// <inheritdoc />` instead, to avoid the two from getting out of sync.